Keep a fallback for the is_member_pointer builtins missing in older g++.

--- a/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/__type_traits/is_member_pointer.h
+++ b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/__type_traits/is_member_pointer.h
@@ -11,6 +11,9 @@
 
 #include <__config>
 #include <__type_traits/integral_constant.h>
+#include <__type_traits/is_function.h>
+#include <__type_traits/remove_cv.h>
+#include <cstddef>
 
 #if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
 #  pragma GCC system_header
@@ -18,6 +21,8 @@
 
 _LIBCPP_BEGIN_NAMESPACE_STD
 
+#if __has_builtin(__is_member_pointer) && __has_builtin(__is_member_object_pointer) && __has_builtin(__is_member_function_pointer)
+
 template <class _Tp>
 struct _LIBCPP_NO_SPECIALIZATIONS is_member_pointer : _BoolConstant<__is_member_pointer(_Tp)> {};
 
@@ -40,6 +45,46 @@
 _LIBCPP_NO_SPECIALIZATIONS inline constexpr bool is_member_function_pointer_v = __is_member_function_pointer(_Tp);
 #endif
 
+#else // __has_builtin(__is_member_pointer) && etc.
+
+template <class _Tp>
+struct __libcpp_is_member_pointer {
+  enum { __is_member = false, __is_func = false, __is_obj = false };
+};
+template <class _Tp, class _Up>
+struct __libcpp_is_member_pointer<_Tp _Up::*> {
+  enum {
+    __is_member = true,
+    __is_func   = is_function<_Tp>::value,
+    __is_obj    = !__is_func,
+  };
+};
+
+template <class _Tp>
+struct _LIBCPP_NO_SPECIALIZATIONS is_member_pointer
+    : public _BoolConstant<__libcpp_is_member_pointer<__remove_cv_t<_Tp> >::__is_member> {};
+
+template <class _Tp>
+struct _LIBCPP_NO_SPECIALIZATIONS is_member_object_pointer
+    : public _BoolConstant<__libcpp_is_member_pointer<__remove_cv_t<_Tp> >::__is_obj> {};
+
+template <class _Tp>
+struct _LIBCPP_NO_SPECIALIZATIONS is_member_function_pointer
+    : public _BoolConstant<__libcpp_is_member_pointer<__remove_cv_t<_Tp> >::__is_func> {};
+
+#if _LIBCPP_STD_VER >= 17
+template <class _Tp>
+_LIBCPP_NO_SPECIALIZATIONS inline constexpr bool is_member_pointer_v = is_member_pointer<_Tp>::value;
+
+template <class _Tp>
+_LIBCPP_NO_SPECIALIZATIONS inline constexpr bool is_member_object_pointer_v = is_member_object_pointer<_Tp>::value;
+
+template <class _Tp>
+_LIBCPP_NO_SPECIALIZATIONS inline constexpr bool is_member_function_pointer_v = is_member_function_pointer<_Tp>::value;
+# endif
+
+#endif // __has_builtin(__is_member_pointer) && etc.
+
 _LIBCPP_END_NAMESPACE_STD
 
 #endif // _LIBCPP___TYPE_TRAITS_IS_MEMBER_POINTER_H
